src/Symfony/Component/Security/Core/Authentication/Token/SwitchUserToken.php

```diff
@@ -35,7 +35,7 @@ public function __construct($user, $credentials, string $providerKey, array $rol
         $this->originalToken = $originalToken;
     }
 
-    public function getOriginalToken(): TokenInterface
+    public function getOriginalToken(): ?TokenInterface
```

Comment:
I don't understand why `getOriginalToken()` should be able to return `null`. How would you end the impersonation if you don't know the original token that you need to switch back to?

Reply:
No need for switch back when impersonating in a different firewall. AdminUser is still logged as itself in the admin context, but also logged as impersonated user in the other firewall. 

AFAIK it's impossible to access the original token (of admin context) from the front context. And I think it's useless in a cross-firewall context. Switching back is relevant in a single firewall context because authentication only handles one token at a time.

I want to check the impersonation status to display a warning, not to provide a switch back link.

Reply:
My point is that the `SwitchUserListener` relies on the `SwitchUserToken` to return the original token and not `null` here. As soon as we allow the original token to be `null` people will end up in situations where that's the case and thus break the built-in feature. We shouldn't make it easier than necessary to shoot yourself in the foot.

I would rather like to see two questions answered before considering introducing this interface: Do you really need two different firewalls? And if so, why is it an actual issue to store the original token?

Reply:
> My point is that the `SwitchUserListener` relies on the `SwitchUserToken` to return the original token and not `null` here. As soon as we allow the original token to be `null` people will end up in situations where that's the case and thus break the built-in feature. We shouldn't make it easier than necessary to shoot yourself in the foot.

IMO, SwitchUserListener should only handle existing SwitchUserToken case, not all impersonation cases. So maybe the interface would benefit being renamed to `ImpersonatedTokenInterface`.

> Do you really need two different firewalls?

I think it's possible to merge my firewalls into one. But, from a security point of view, why providing security splitted firewalls if we need to merge them into one ? I prefer having one firewall per application context instead of relying on roles to split contexts access. IMO, roles are great for fine access definitions inside a context, not really for granting admin/extranet/front contexts.

> And if so, why is it an actual issue to store the original token?

(AFAIK) It's not always possible, and a non-sense across multiple firewalls. Original token is only required for the "switch back" feature, because only one token may be taken in account in a single firewall. Within multiple firewalls context, each firewall handles its dedicated token.

Reply:
Anyway, thanks for your time trying to decrypt my need/intention ;-)